### PR TITLE
Generic downstream eval

### PR DESCRIPTION
**Pull request by mrudat-iais**
# What does this PR do?
We can now convert a modalities model into a HF model. This model can then be evaluated in eval harness. 
* (**main**) We now have a `convert_pytorch_to_hf_checkpoint` endpoint that takes the `model_name`, `model_config_path` and `output_dir` and creates HF files that can then be loaded with `AutoModel.from_pretrained`
* (**tests**) we have a test that shows the functionality [tests/checkpointing/test_checkpoint_conversion.py](https://github.com/Modalities/modalities/blob/019b8074f79efefa484c580824b4bf9b60ace5da/tests/checkpointing/test_checkpoint_conversion.py)

## General changes
* none besides the above
* PR feedback addressed and implemented

## Breaking Changes
* no breaking changes

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue / enhancement in isolation
- [x] I have merged main into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [ ] I have run a sample config for model training
- [x] I have fixed all failing tests (`python tests/tests.py`)

### CODE REVIEW

**Review by le1nux — COMMENTED** [review]
Really good work and awesome that you guys found a generic solution to it! :-) Thanks a lot for digging deep into the huggingface source code!

I left a few remarks here and there that you should address but overall the solution is great!

**Review by le1nux — CHANGES_REQUESTED** [review]
Really good work and awesome that you guys found a generic solution to it! :-) Thanks a lot for digging deep into the huggingface source code!

I left a few remarks here and there that you should address but overall the solution is great!

**Review by le1nux — COMMENTED** [review]
Really good work and awesome that you guys found a generic solution to it! :-) Thanks a lot for digging deep into the huggingface source code!

I left a few remarks here and there that you should address but overall the solution is great!

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from modalities.models.utils import get_model_from_config
+
+
+class HFAdapterConfig(PretrainedConfig):
+    model_type = "modalities"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if self.config:
+            self.convert_posixpath_to_str(self.config)
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:21]: We added an explanation, but decided to omit the example since it's not a complex operation

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        Recursively iterate over the dictionary and convert PosixPath values to strings.
+        """
+        for key, value in d.items():
+            if isinstance(value, PosixPath):
+                d[key] = str(value)
+            elif isinstance(value, dict):
+                self.convert_posixpath_to_str(value)
+            elif isinstance(value, list):
+                d[key] = [str(item) if isinstance(item, PosixPath) else item for item in value]
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:40]: These are `PosixPaths`s from our config

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+class HFAdapterConfig(PretrainedConfig):
+    model_type = "modalities"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if self.config:
+            self.convert_posixpath_to_str(self.config)
+
+    def to_json_string(self, use_diff: bool = True) -> str:
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:23]: The override decorator is only supported by python 3.12 or newer, but in our pyproject.toml we actually require python < 3.12

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            elif isinstance(value, list):
+                d[key] = [str(item) if isinstance(item, PosixPath) else item for item in value]
+
+
+class HFAdapter(PreTrainedModel):
+    config_class = HFAdapterConfig
+
+    def __init__(self, config: HFAdapterConfig, model: Optional[nn.Module] = None, *inputs, **kwargs):
+        super().__init__(config)
+        if not model:
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:48]: We removed the model-parameter and now always instantiate the model internally

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        output_hidden_states: Optional[bool] = False,
+    ):
+        if output_attentions or output_hidden_states:
+            raise NotImplementedError
+        model_input = {"input_ids": input_ids, "attention_mask": attention_mask}
+        model_forward_output: Dict[str, torch.Tensor] = self.model.forward(model_input)
+        if return_dict:
+            return ModalitiesModelOutput(**model_forward_output)
+        else:
+            return model_forward_output[self.model.prediction_key]
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:68]: We think in our case it is necessary that the model has a prediction key. Else how should we know which entry to access in model_forward_output? We add therefore an explicit assertion to ensure that model.prediction_key exists

File: tests/checkpointing/test_checkpoint_conversion.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    return test_tensor
+
+
+def test_hf_and_pytorch_models_are_the_same_after_init(hf_model, pytorch_model, checkpoint_conversion):
+    assert hf_model.dtype == pytorch_model.lm_head.weight.dtype
+    assert os.listdir(checkpoint_conversion.output_hf_checkpoint_dir)
+
+
+def test_models_before_and_after_conversion_produce_same_output(
+    device,
```
[the remark this answers is not in the document]
Comment (rrutmann) [reply tests/checkpointing/test_checkpoint_conversion.py:109]: Do we really need type hints for test fixtures? Isn't it enough to have the return type for each fixture?

File: src/modalities/models/huggingface_adapters/hf_adapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:25]: true, I also saw this today. We can keep it for later version of modalities. :-)

File: src/modalities/models/huggingface_adapters/hf_adapter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:74]: The issue is that `NNModel`does not have the attribute `prediction_key` and cannot assume that every inherited model has it. So by "contract" of the class definitions, it is totally fine that a model does not have a `prediction_key` and it would throw an error in the HFAdapter at runtime. 

Additionally, for different use cases other prediction keys might be expected. E.g., argmax vs softmax outputs are both predictions that we could use in different scenarios. 

I would therefore specify the needed prediction key for the HFAdpater in the config yaml  by referencing the key.

File: tests/checkpointing/test_checkpoint_conversion.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/checkpointing/test_checkpoint_conversion.py]: I think it makes it more practical when you also add the type arguments to the fixture arguments. This way autocomplete for instance works and static code analysis such as type checks also work.

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        for key, value in d.items():
+            if isinstance(value, PosixPath):
+                d[key] = str(value)
+            elif isinstance(value, dict):
+                self._convert_posixpath_to_str(value)
+            elif isinstance(value, list):
+                for item in value:
+                    if isinstance(item, PosixPath):
+                        d[key] = str(item)
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:44]: I think this is a bug. 
here we replace a list of Posixpaths and posibly other items to a single posixpath string. Is this wanted behavior?

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                    elif isinstance(item, dict):
+                        self._convert_posixpath_to_str(item)
+                    else:
+                        d[key] = item
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:48]: same issue as above

File: src/modalities/models/huggingface_adapters/hf_adapter.py (hunk 7 above)
Comment (le1nux) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:44]: We should cover this method with unit testing.

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+
+class HFModelAdapter(PreTrainedModel):
+    config_class = HFModelAdapterConfig
+
+    def __init__(self, config: HFModelAdapterConfig, *inputs, **kwargs):
+        super().__init__(config, *inputs, **kwargs)
+        self.model: NNModel = get_model_from_config(config.config, model_type="checkpointed_model")
+        assert hasattr(self.model, "prediction_key"), "Missing entry model.prediction_key in config"
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:57]: asserts are often deactivated in production code and it is best practice to raise exceptions instead. 

Following the discusion from below, why don't we pass the prediction key to the constructor of HFModelAdapter, instead? We could prevent this "hack" of checking that a model has a prediction key.

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        return {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+        }
+
+
+@dataclass
+class ModalitiesModelOutput(ModelOutput):
+    logits: torch.FloatTensor = None
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:91]:
```suggestion
    logits: Optional[torch.FloatTensor] = None
```

also applies to the other attribtes below

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            raise NotImplementedError
+        model_input = {"input_ids": input_ids, "attention_mask": attention_mask}
+        model_forward_output: Dict[str, torch.Tensor] = self.model.forward(model_input)
+        if return_dict:
+            return ModalitiesModelOutput(**model_forward_output)
+        else:
+            return model_forward_output[self.model.prediction_key]
+
+    def prepare_inputs_for_generation(
+        self, input_ids: torch.LongTensor, attention_mask=None, **kwargs
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:77]: type missing for `attention_mask``

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        attention_mask: Optional[torch.Tensor] = None,
+        return_dict: Optional[bool] = False,
+        output_attentions: Optional[bool] = False,
+        output_hidden_states: Optional[bool] = False,
+    ):
+        if output_attentions or output_hidden_states:
+            raise NotImplementedError
+        model_input = {"input_ids": input_ids, "attention_mask": attention_mask}
+        model_forward_output: Dict[str, torch.Tensor] = self.model.forward(model_input)
+        if return_dict:
```
Comment (le1nux) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:71]: not sure if this is a bug, but if the return_dict is True we return `ModalitiesModelOutput` and the dict otherwise. Shouldn't it be the other way around?

File: src/modalities/models/utils.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+from pydantic import BaseModel
+
+from modalities.config.component_factory import ComponentFactory
+from modalities.config.pydanctic_if_types import PydanticPytorchModuleType
+from modalities.registry.components import COMPONENTS
+from modalities.registry.registry import Registry
+    
+    
+def get_model_from_config(config: dict, model_type: str):
```
Comment (le1nux) [inline src/modalities/models/utils.py:9]: We should make `model_type` an Enum

File: tests/checkpointing/test_checkpoint_conversion.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
+import os
+from pathlib import Path
```
Comment (le1nux) [inline tests/checkpointing/test_checkpoint_conversion.py:2]: type hints are missing basically everywhere in the tests. That makes it difficult to verify the correctness of the tests.

File: tests/checkpointing/test_checkpoint_conversion.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def test_models_before_and_after_conversion_are_equal(
+    pytorch_model,
+    hf_model,
+    hf_model_from_checkpoint,
+):
+    for p1, p2, p3 in zip(hf_model.parameters(), pytorch_model.parameters(), hf_model_from_checkpoint.parameters()):
+        assert torch.equal(p1, p2)
+        assert torch.equal(p1, p3)
```
Comment (le1nux) [inline tests/checkpointing/test_checkpoint_conversion.py:101]: Another test would be great that the outputs  of  `pytorch_model`, `hf_model` and `hf_model_from_checkpoint` are the same for a given random input batch.

**Review by le1nux — COMMENTED** [review]
The PR is almost in a mergable state. I found one potential bug with the PosixPath to string conversion. And some minor things that I think should be addressed before merging. Other than that it's in a great state and we should be able to merge after including the requested changes and another person has reviewed it :-)

File: tests/checkpointing/test_checkpoint_conversion.py (hunk 15 above)
Comment (rrutmann) [reply tests/checkpointing/test_checkpoint_conversion.py:137]: We had this test before, but isn't this implied if all parameters are the same?

File: src/modalities/models/huggingface_adapters/hf_adapter.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        return_dict: Optional[bool] = False,
+        output_attentions: Optional[bool] = False,
+        output_hidden_states: Optional[bool] = False,
+    ):
+        if output_attentions or output_hidden_states:
```
Comment (mali-git) [inline src/modalities/models/huggingface_adapters/hf_adapter.py:66]: Please add a comment describing this case.

File: src/modalities/models/huggingface_adapters/hf_adapter.py (hunk 12 above)
Comment (mali-git) [reply src/modalities/models/huggingface_adapters/hf_adapter.py:70]: This is also not clear to me. 
@rrutmann, can you please explain this?

File: src/modalities/models/utils.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+from modalities.config.component_factory import ComponentFactory
+from modalities.config.pydanctic_if_types import PydanticPytorchModuleType
+from modalities.registry.components import COMPONENTS
+from modalities.registry.registry import Registry
+
+class ModelTypeEnum(Enum):
+    MODEL = "model"
+    CHECKPOINTED_MODEL = "checkpointed_model"
+    
+def get_model_from_config(config: dict, model_type: ModelTypeEnum):
```
Comment (mali-git) [inline src/modalities/models/utils.py:13]: `dict `is not a type annotation. Please use `Dict` from typing and the type annotations for key an value.

File: tests/models/test_hf_adapter.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+from modalities.models.huggingface_adapters.hf_adapter import HFModelAdapterConfig
+
+
+@pytest.fixture()
+def hf_model_adapter_config() -> HFModelAdapterConfig:
+    return HFModelAdapterConfig(config={})
+
+
+def test___convert_posixpath_to_str(hf_model_adapter_config: HFModelAdapterConfig):
```
Comment (mali-git) [inline tests/models/test_hf_adapter.py:13]: To many underscore characters

**Review by mali-git — CHANGES_REQUESTED** [review]
